drivers: pwm: generate pwm with nxp flexio IP

sumitbatra-nxp opened the pull request:
This pwm driver can be used for both mcux and S32 SoCs which have the flexio IP.
The requested pwm channels from dt are automatically allocated by the flexio misc driver based on the timers available.

mmahadevan108 commented on dts/arm/nxp/nxp_s32k344_m7.dtsi:
```diff
@@ -828,6 +828,20 @@
+			clocks = <&clock NXP_S32_FLEXIO0_CLK>;
+			status = "disabled";
+
+			pwm {
```
To be consistent with label names used for spi, please change this to flexio0_pwm

mmahadevan108 commented on drivers/pwm/Kconfig.mcux_flexio:
```diff
@@ -0,0 +1,11 @@
+# Copyright 2024 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+config PWM_MCUX_FLEXIO
```
Please change PWM_MCUX_FLEXIO to PWM_NXP_FLEXIO

mmahadevan108 commented on drivers/pwm/Kconfig.mcux_flexio:
```diff
@@ -0,0 +1,11 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config PWM_MCUX_FLEXIO
+	bool "MCUX Flexio PWM driver"
```
MCUX to NXP

mmahadevan108 commented on boards/nxp/mr_canhubk3/mr_canhubk3-pinctrl.dtsi:
Please update the documentation to indicate FlexIO support and PWM support over FlexIO.

manuargue reviewed (changes requested):
LGTM, just minor nits. @sumitbatra-nxp was this patch tested with any pwm test/sample?

sumitbatra-nxp commented:
> LGTM, just minor nits. @sumitbatra-nxp was this patch tested with any pwm test/sample?

Both pwm loopback (just that I commented out the capture) and pwm api (where I made it to use both  pwm-0 and pwm-1 alias

sumitbatra-nxp commented on drivers/pwm/pwm_mcux_flexio.c:
```diff
@@ -0,0 +1,386 @@
+#include <zephyr/drivers/misc/nxp_flexio/nxp_flexio.h>
+
+
+LOG_MODULE_REGISTER(pwm_mcux_flexio, CONFIG_PWM_LOG_LEVEL);
```
This was my dilemma, since it is for both mcux and s32

danieldegrasse commented on drivers/pwm/Kconfig.nxp_flexio:
```diff
@@ -0,0 +1,11 @@
+
+config PWM_NXP_FLEXIO
+	bool "NXP Flexio PWM driver"
+	default n
```
Why is this `default n`? Typically drivers are `default y`, so they are enabled whenever the relevant API class is enabled and a device with the appropriate compatible is enabled (which will set the `DT_HAS_xxx_ENABLED` Kconfig)

sumitbatra-nxp replied:
@danieldegrasse 
Twister build target for board twr_ke18f was failing for me, if I enabled flexio pwm driver by default 
west build -p always -b twr_ke18f  tests/drivers/build_all/pwm/
 
**Reason is**: KE1x family SoCs does have PWM support but doesn't have PINOUTD or PINOUTE registers, but these registers are required by S32 and MCX and i.Mx SoCs. 
So driver APIs that I added under FSL_FEATURE_FLEXIO_HAS_PIN_REGISTER are undefined for ke18f build target.

**Solution**: I can check for FSL_FEATURE_FLEXIO_HAS_PIN_REGISTER for Pin related API calls in pwm_nxp_flexio.c

anangl commented on drivers/pwm/pwm_nxp_flexio.c:
```diff
@@ -0,0 +1,374 @@
+	};
+
+#define FLEXIO_PWM_TIMER_INDEX_INIT(n)								\
+	uint8_t flexio_pwm_##n##_timer_index[ARRAY_SIZE(flexio_pwm_##n##_init)];
```
Why is it not `static`?

sumitbatra-nxp replied:
done

sumitbatra-nxp replied:
done

anangl commented on drivers/pwm/pwm_nxp_flexio.c:
```diff
@@ -0,0 +1,374 @@
+	},
+
+#define FLEXIO_PWM_PULSE_GEN_CONFIG(n)								\
+	pwm_nxp_flexio_channel_config_type flexio_pwm_##n##_init[] = {				\
+		DT_INST_FOREACH_CHILD_STATUS_OKAY(n, _FLEXIO_PWM_PULSE_GEN_CONFIG)		\
+	};											\
+	const struct pwm_nxp_flexio_pulse_info flexio_pwm_##n##_info = {			\
```
And these two as well.

sumitbatra-nxp replied:
done

anangl commented on drivers/pwm/pwm_nxp_flexio.c:
```diff
@@ -0,0 +1,374 @@
+	};
+
+#define FLEXIO_PWM_PULSE_GEN_GET_CONFIG(n)							\
+	.pulse_info = (struct pwm_nxp_flexio_pulse_info *)&flexio_pwm_##n##_info,
```
Instead of this typecasting you should rather define the `pulse_info` field as `const pwm_nxp_flexio_channel_config_type *`.

sumitbatra-nxp replied:
pwm_nxp_flexio_pulse_info has 2 fields .pwm_pulse_channels (which is the number of pwm channels) and 
.pwm_info which is of type pwm_nxp_flexio_channel_config_type is an array with channel configurations for each channel.

Why do you want me to pulse_info to type channel_config ?

sumitbatra-nxp replied:
I have removed the typecasting and made pulse_info field as const.

anangl commented on drivers/pwm/pwm_nxp_flexio.c:
```diff
@@ -0,0 +1,374 @@
+		}										\
+	};
+
+#define FLEXIO_PWM_PULSE_GEN_GET_CONFIG(n)							\
+	.pulse_info = (struct pwm_nxp_flexio_pulse_info *)&flexio_pwm_##n##_info,
```
Do you need to have this as a macro? Why not assign `.pulse_info` directly in `PWM_NXP_FLEXIO_PWM_INIT()`?

sumitbatra-nxp replied:
I just want to show pwm channels separately, that's it..

anangl commented on drivers/pwm/pwm_nxp_flexio.c:
```diff
@@ -0,0 +1,374 @@
+	};											\
+	const struct pwm_nxp_flexio_pulse_info flexio_pwm_##n##_info = {			\
+		.pwm_pulse_channels = ARRAY_SIZE(flexio_pwm_##n##_init),			\
+		.pwm_info = (pwm_nxp_flexio_channel_config_type *)flexio_pwm_##n##_init,	\
```
nit: this typecasting is redundant.

sumitbatra-nxp replied:
If I remove this typecasting the compiler says-
warning: initialization discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
  328 |                 .pwm_info = flexio_pwm_##n##_init,

sumitbatra-nxp replied:
removed !

dleach02 commented:
@sumitbatra-nxp, please address the CI failures, merge conflicts, and reviewer comments so we can close on this PR

manuargue reviewed (changes requested):
pls update board's docs to mention FlexIO PWM support https://github.com/zephyrproject-rtos/zephyr/blob/3c8e204bcf19d87eba22aaed63647f10f8f8c6f4/boards/nxp/mr_canhubk3/doc/index.rst?plain=1#L46

manuargue reviewed (dismissed):
all my comments addressed

danieldegrasse commented on drivers/pwm/pwm_nxp_flexio.c:
```diff
@@ -0,0 +1,374 @@
+};
+
+enum pwm_nxp_flexio_prescaler {
+	/* Decrement counter on Flexio clock */
```
Nit- please be consistent between using `@brief` style documentation and simple comments. I'd prefer comments here, but either method is fine provided it is consistent

sumitbatra-nxp replied:
Removed @brief

danieldegrasse commented on dts/bindings/pwm/nxp,flexio-pwm.yaml:
```diff
@@ -0,0 +1,64 @@
+        pin-id should be populated with number 'n' specified in the
+        FXIO_D<n> data pin of your board's schematic.
+
+    polarity:
```
We could instead define this as a string enum, like so:
```
    polarity:
      type: string
      enum:
        - "active-low"
        - "active-high"
      description: |
        Output polarity for PWM channel.
```

`DT_ENUM_IDX` could then be used to convert the polarity selection to an integer value

sumitbatra-nxp replied:
Done !!

dleach02 commented:
@sumitbatra-nxp I updated your west.yml to align with the merging of the HAL side

henrikbrixandersen commented on boards/nxp/mr_canhubk3/mr_canhubk3.dts:
```diff
@@ -616,6 +617,33 @@
+		pinctrl-names = "default";
+		status = "okay";
+
+		pwm_0 {
+			pin-id = <19>;
+			period = <400>;
+			duty-cycle = <200>;
+			prescaler = <0>;
+			polarity = "active-high";
+		};
```
This is not typically how we configure PWM outputs in devicetree? Configuration is done from the PWM client node, using a `*pwms = <&flexio0_pwm ...>` property.

sumitbatra-nxp replied:
@henrikbrixandersen 
Shall I remove only polarity or both polarity and period from the device tree properties (as both polarity and period are a part of pwm-cells)

@manuargue .. 
There is no constraint to set the polarity at the init time in flexio PWM mode.
But if I need to remove period as well from the device tree then I will get a low flat pwm line initially and will definitely need a pwm client entry in my pwm app overlay like this -

 pwm_<app>_0 {
                compatible = "test-pwm-<app>";
                pwms = <&flexio0_pwm 0 0 PWM_POLARITY_NORMAL>,
                        <&flexio0_pwm 1 0 PWM_POLARITY_NORMAL>;
        };

and then use macros like - DT_PWMS_PERIOD_BY_IDX, DT_PWMS_FLAGS_BY_IDX in the driver to get the period, polarity at init time

sumitbatra-nxp replied:
@manuargue .. So is it ok if I don't make any overlay entries in a pwm app (samples/test) for this driver.. ?

henrikbrixandersen replied:
> Shall I remove only polarity or both polarity and period from the device tree properties (as both polarity and period are a part of pwm-cells)

I would remove both polarity, duty-cycle and period.

sumitbatra-nxp replied:
@henrikbrixandersen .. I understand that pwm-cells has period and flags (polarity and more), so there is no reason to have duplicate properties but why duty cycle needs to be removed ?

henrikbrixandersen replied:
It is up to the PWM client to specify the duty cycle as well as the other properties. 

I suggest taking a look at how all the other PWM drivers operate.

sumitbatra-nxp replied:
> > So is it ok if I don't make any overlay entries for pwm client nodes in a pwm app (samples/test) for this driver.. ?
> 
> Ideally we should have something to build or test this driver.

@manuargue .. 
tests/drivers/pwm/pwm_api can be used to operate this driver as flexio0_pwm node is pwm-1 alias
But for depicting pwms client node for applications -
I cannot use - tests/drivers/pwm/pwm_loopback as flexio doesn't have a pwm input mode like emios SAIC
So, shall I create an overlay in boards/nxp/mr_canhubk3 or do you suggest any other existing application ?

sumitbatra-nxp replied:
So I have not added anything on the application side, but have removed period, duty_cycle and polarity from the device tree and bindings as suggested by @henrikbrixandersen
But have hardcoded these values for init time in the driver.

henrikbrixandersen replied:
By default, no PWM signal should be produced until `pwm_set_cycles_t` for the given driver is called. You cannot have a default PWM period and duty-cycle upon driver initialisation.

sumitbatra-nxp replied:
Removed default PWM period and duty-cycle upon driver initialization..

dleach02 commented:
@henrikbrixandersen did @sumitbatra-nxp address your change request?

sumitbatra-nxp commented on drivers/pwm/pwm_nxp_flexio.c:
```diff
@@ -0,0 +1,361 @@
+
+		/* Enable the pin out for the selected timer */
+		timerConfig.pinConfig = FLEXIO_PWM_TIMER_PIN_OUTPUT_ENABLE;
+		timerConfig.pinPolarity = pwm_info->polarity;
```
The initial polarity setting in the HAL/low level driver is getting set in line #267
So this assignment is not required now. Will remove it.

sumitbatra-nxp replied:
Set to a default value

sumitbatra-nxp commented on dts/bindings/pwm/nxp,flexio-pwm.yaml:
```diff
@@ -0,0 +1,47 @@
+        pin-id should be populated with number 'n' specified in the
+        FXIO_D<n> data pin of your board's schematic.
+
+    prescaler:
+      type: int
+      description: |
+        The clock divider for internal counter prescaler.
+        - 0: Divide by 1
+        - 4: Divide by 16
+        - 5: Divide by 256
+      enum: [0, 4, 5]
```
Flexio IP's Timer configuration register has some bits which configures the source of the timer decrement and the source of the shift clock. The bits required for PWM configuration can be configured as 0, 4 or 5. 
My primary concern is to configure the IP register  with that. That's why I have considered these values.

manuargue replied:
> I'd prefer for the enum to be [1, 16, 256] and resolve that to the [0, 4, 5] at compile-time

I agree with this. By doing that we can get rid of `pwm_nxp_flexio_get_prescaler_div()` and resolve the value with `_CONCAT(FLEXIO_PWM_CLK_DIV_, DT_INST_PROP(n, prescaler))`

sumitbatra-nxp replied:
Thanks @manuargue 
Made this change, but since there is no direct logic to convert - 0,4,5 to 1,16,256 so I had to add another config member prescaler_div.
A better suggestion is most welcome..

sumitbatra-nxp commented on drivers/pwm/pwm_nxp_flexio.c:
```diff
@@ -0,0 +1,361 @@
+	/** Pwm period in ticks */
+	uint16_t period;
+	/** Pwm duty cycle in ticks */
+	uint16_t duty_cycle;
```
'period' is used in pwm_nxp_flexio_get_cycles_per_sec as well, so I'll keep an array of period in pwm_nxp_flexio_data

sumitbatra-nxp replied:
Done !!

sumitbatra-nxp commented on drivers/pwm/pwm_nxp_flexio.c:
```diff
@@ -0,0 +1,361 @@
+	FLEXIO_PWM_TIMER_PIN_OUTPUT_ENABLE   = 0x03U
+};
+
+typedef struct {
```
Removed the typedef

danieldegrasse commented on west.yml:
Please drop this change to the `west.yml`. This currently is rolling the NXP HAL back to an older SHA, which is the reason for the KE17Z related build failures in this PR

sumitbatra-nxp replied:
Dropped !!

sumitbatra-nxp committed:
drivers: pwm: generate pwm with nxp flexio IP

This driver can be used for both mcux and
s32k series SoCs which have flexio IP.
PWM channel is automatically allocated by
flexio driver based on the available timers.

Signed-off-by: Sumit Batra <sumit.batra@nxp.com>

dleach02 commented:
@henrikbrixandersen did @sumitbatra-nxp address your change request?
